### ISSUE TITLE
```DataCollection``` functions should be split and named more clearly

### ISSUE THREAD

**Issue by c-prather** [parthenon-hpc-lab/parthenon#906]
I was recently stung by a bug in which I wrote the following (paraphrased):

```c++
if (stage == 1)
    pmb->meshblock_data.Add("preserve", base);
```

And then expected that the `"preserve"` member of the `DataCollection` contained the values from `base`.  In the first step, or the first step for a new meshblock, or after a restart, it *would* contain the correct data.  However, it is not reliably updated during normal operation, since the `meshblock_data` object is re-used, and thus would not allocate the `"preserve"` container, and thus not copy into it.

This is my bug, to be clear.  The behavior is consistent enough with the naming `Add`.  However, given it takes an argument to copy from, I could be forgiven for assuming that it always resulted in a copy.

This doesn't seem like useful behavior to add to the allocation step.  Instead, I'd advocate splitting the function in two: `Add("name")`, which adds & allocates variables present in `base` without any copying (and, if you insist, also `AddLike("name", source)` which would do the same for an arbitrary MBD).  This could allocate the memory without ever copying anything, since presumably everyone fills the resulting container for themselves rather than relying on current behavior.  Then a separate `Copy` which would always perform the copy, allocating the target if needed.

Happy to write the implementation if downstreams are happy with the (quite thoroughly breaking) change.

Two other incidentals which would affect the eventual PR though:
1. I'm not sure how `GetOrAdd` fits into this.  Isn't that already how `Add` should behave?
2. Calling the `MeshBlock` member `meshblock_data` when it is in fact a `DataCollection` of `MeshBlockData` is quite confusing.
3. Is there any reason the source argument is a pointer instead of a string?  The `DataCollection` can just do the lookup instead of the user, yeah?

**Comment by Yurlungur** [comment]
I don't think this interface has been particularly carefully thought through and it might be good to be more deliberate. A few things:
1. `GetOrAdd` is specifically for `MeshData` and we added it to have the relevant `MeshBlock` data available on each block and tied to the higher-level `MeshData`
2. `meshblock_data` is perfectly reasonable as a name for a `DataCollection`. It's plural. :P
3. I don't object to *adding* a string argument overload, but I prefer the pointer as it is less error prone than hand-typed strings. I'd like to move away from hand-typed strings as much as possible.

**Comment by c-prather** [comment]
1. If it's only for `MeshData`, but does the same thing, why not make it another overload of `Add` (or actually, maybe name both `GetOrAdd`)?  I think I could pull that off.
2. I submit `stages`, given we already call its keys `stage_names` (sure, not all the keys one might add, but it gets the main point across)
3. I get that argument; I'm happy having both.  I think I'll be less prone to error with the string since I can use literals, but I could see the typing being useful as things get more complex.

**Comment by Yurlungur** [comment]
`GetOrAdd` works a little differently than `Add`---it's designed to add `MeshBlockData` with the same name as the `MeshData` but with `_str(i)` for each `MeshBlock` in the `MeshData`
